contrib/datadog/tracing/interceptor.go

```diff
@@ -0,0 +1,200 @@
+	return interceptor.TracerOptions{
+		SpanContextKey: activeSpanContextKey,
+		HeaderKey:      headerKey,
+	}
+}
+
+func (tracerImpl) UnmarshalSpan(m map[string]string) (interceptor.TracerSpanRef, error) {
+	var carrier tracer.TextMapCarrier = m
+	ctx, err := tracer.Extract(carrier)
+	if err != nil {
+		if errors.Is(err, tracer.ErrSpanContextNotFound) {
+			// If there is no span, return nothing, but don't error out. This is
+			// a legitimate place where a span does not exist in the headers
+			return nil, nil
+		}
+		return nil, err
+	}
+	return &tracerSpanCtx{ctx}, nil
+}
+
+func (tracerImpl) MarshalSpan(span interceptor.TracerSpan) (map[string]string, error) {
+	carrier := tracer.TextMapCarrier{}
+	if err := tracer.Inject(span.(*tracerSpan).Context(), carrier); err != nil {
+		return nil, err
+	}
+	return carrier, nil
+}
+
+func (tracerImpl) SpanFromContext(ctx context.Context) interceptor.TracerSpan {
+	span, ok := tracer.SpanFromContext(ctx)
+	if !ok {
+		return nil
+	}
+	return &tracerSpan{Span: span}
+}
+
+func (tracerImpl) ContextWithSpan(ctx context.Context, span interceptor.TracerSpan) context.Context {
+	return tracer.ContextWithSpan(ctx, span.(*tracerSpan).Span)
+}
+
+func (tracerImpl) StartSpan(options *interceptor.TracerStartSpanOptions) (interceptor.TracerSpan, error) {
+	startOpts := []tracer.StartSpanOption{
+		tracer.ResourceName(options.Name),
+		tracer.StartTime(options.Time),
+	}
+
+	// Set a deterministic span ID for workflows which are long-running and cross process boundaries
+	if options.Operation == "RunWorkflow" {
+		rid, ok := options.Tags["temporalRunID"]
+		if !ok {
+			return nil, fmt.Errorf("missing required tag `temporalRunID` on `RunWorkflow` span")
+		}
+		h := fnv.New64()
+		// Write() always writes all bytes and never fails; the count and error result are for implementing io.Writer.
+		_, _ = h.Write([]byte(fmt.Sprintf("workflow:%s", rid)))
+		startOpts = append(startOpts, tracer.WithSpanID(h.Sum64()))
+	}
+
+	// Add parent span to start options
+	var parent ddtrace.SpanContext
+	switch opParent := options.Parent.(type) {
+	case nil:
+	case *tracerSpan:
+		parent = opParent.Context()
+	case *tracerSpanCtx:
+		parent = opParent
+	default:
+		return nil, fmt.Errorf("unrecognized parent type %T", opParent)
+	}
+	if parent != nil {
+		// Convert the parent context into a full trace (required, otherwise
+		// the implementation of StartSpan() will NOT respect ChildOf()
+		// The runtime type of parentTrace will be tracer.spanContext (note the
+		// starting lowercase on "spanContext", that's an internal struct)
+		parentTrace, err := tracer.Extract(newSpanContextReader(parent))
+		if err != nil {
+			return nil, err
+		}
+		startOpts = append(startOpts, tracer.ChildOf(parentTrace))
+	}
+
+	// Add tags to start options
+	for k, v := range options.Tags {
+		// Display Temporal tags in a nested group in Datadog APM
+		tagKey := strings.Replace(k, "temporal", "temporal.", 1)
```

Comment:
We do this because datadog groups by a "dot". This way all tags that share the "temporal." prefix are grouped in the interface. It could break if temporal ever adds its own "grouping"